src/social/api.ts

```diff
@@ -1,124 +1,222 @@
+     *
+     * @returns the downloaded PostCommandLog
+     */
+    downloadPostCommandLog: (since?: Swarm.Epoch) => Promise<PostCommandLog>;
+}
 
-const timestampCompare = (a: PostCommand, b: PostCommand) => {
-    return a.timestamp - b.timestamp;
-};
+/**
+ * This interface must be implemented by storages that intends to
+ * store RecentPostFeeds.
+ */
+export interface RecentPostFeedStorage {
+    /**
+     * Uploads (and updates) the RecentPostFeed
+     *
+     * @param postCommandLog contains all the updates that is used to calculate
+     * the recent updates
+     * @param recentPostFeed used as a template for creating the feed update
+     *
+     * @returns the updated RecentPostFeed
+     */
+    uploadRecentPostFeed: (postCommandLog: PostCommandLog, recentPostFeed: RecentPostFeed) => Promise<RecentPostFeed>;
+
+    /**
+     * Downloades the RecentPostFeed
+     *
+     * @param url the url of the feed
+     * @param timeout the maximum time to wait to download the feed, if exceeded an
+     * exception is thrown
+     *
+     * @returns the downloaded RecentPostFeed
+     */
+    downloadRecentPostFeed: (url: string, timeout: number) => Promise<RecentPostFeed>;
+}
 
-const sourceCompare = (a: PostCommand, b: PostCommand) => {
-    return a.source.localeCompare(b.source);
-};
+/**
+ * A storage must implement post command log storage and recent
+ * feed storage as well.
+ */
+export type Storage = PostCommandLogStorage & RecentPostFeedStorage;
+
+/**
+ * The updates after a sync is returned in this interface
+ */
+export interface StorageSyncUpdate {
+    /**
+     * The updated and synced PostCommandLog
+     */
+    postCommandLog: PostCommandLog;
+    /**
+     * The updated RecentPostFeed
+     */
+    recentPostFeed: RecentPostFeed;
+    /**
+     * The updated posts
+     */
+    updatedPosts: Post[];
+}
 
-const mergePostCommandLogs = (postCommandLogA: PostCommandLog, postCommandLogB: PostCommandLog): PostCommandLog => {
-    const mergedPostCommandLog = {
-        commands: postCommandLogA.commands.concat(postCommandLogB.commands),
-    };
-    const sortedPostCommandLog = {
-        commands: mergedPostCommandLog.commands.sort((a, b) =>
-            timestampCompare(a, b) || sourceCompare(a, b)),
-    };
-    return sortedPostCommandLog;
-};
+/**
+ * This interface must be implemented by storages which supports syncing.
+ */
+export interface StorageSyncer {
+    /**
+     * Syncs the post command logs between the storage and the local command log.
+     *
+     * @param postCommandLog the local command log
+     * @param recentPostFeed the local recent post feed
+     *
+     * @returns the updates after syncing
+     */
+    sync: (postCommandLog: PostCommandLog, recentPostFeed: RecentPostFeed) => Promise<StorageSyncUpdate>;
+}
 
+/**
+ * Shares a new post and puts it in the command log
+ *
+ * @param post the post to be shared
+ * @param source the unique identifier of a user/device
```

Comment:
are there any privacy/security implications?

Reply:
Yes, with public posts it leaks the data what device you used to create the post. I was thinking of several ideas how this can be mitigated:
- If you know the devices ahead (because there is a pairing process for example) you can use some derivation from the device ids, that has some shared secret involved (e.g. hmac). In this case the device info is not publicly visible and if we add the timestamp to the value to be hashed it will change with every post. The downside is that to track the source we must calculate all the hash of the known device ids to identify the source.
- An alternative is to use the `privateKey` as a symmetric encryption key with the `timestamp` as nonce to encrypt the source. I read somewhere that this may not be a good idea, so further research is needed.
- Or actually we can just use random if we don't really care about which device created the data. The sorting/merging algorithm will then work correctly assuming there are no conflicting ids which may break the invariants, but if the random space is big then this is not an issue.

Reply:
There are 2 things which come to my mind:
- it shouldn't be _the_ device id, or anything which is used for other then our application, because that would definitely mean a privacy leak
- maybe (after integration swarm light-node) we can use something specific to the node. That would satisfy the above requirement, and it's also different from device to device for the same user